include/uc_priv.h

```diff
@@ -60,16 +60,53 @@ typedef bool (*uc_args_int_t)(int intno);
 // some architecture redirect virtual memory to physical memory like Mips
 typedef uint64_t (*uc_mem_redirect_t)(uint64_t address);
 
-
-struct hook_struct {
-    int hook_type; // uc_tracecode_type & uc_tracemem_type
-    uint64_t begin, end;    // range of address to be monitored
-    void *callback; // either uc_cb_tracecode_t or uc_cb_tracemem_t
+struct hook {
+    int type;            // UC_HOOK_*
+    int insn;            // instruction for HOOK_INSN
+    uint64_t begin, end; // only trigger if PC or memory access is in this address (depends on hook type)
+    void *callback;      // a uc_cb_* type
     void *user_data;
 };
 
-// extend memory to keep 32 more hooks each time
-#define HOOK_SIZE 32
+// hook list offsets
+// mirrors the order of uc_hook_type from include/unicorn/unicorn.h
+enum uc_hook_idx {
+    UC_HOOK_INTR_IDX,
+    UC_HOOK_INSN_IDX,
+    UC_HOOK_CODE_IDX,
+    UC_HOOK_BLOCK_IDX,
+    UC_HOOK_MEM_READ_UNMAPPED_IDX,
+    UC_HOOK_MEM_WRITE_UNMAPPED_IDX,
+    UC_HOOK_MEM_FETCH_UNMAPPED_IDX,
+    UC_HOOK_MEM_READ_PROT_IDX,
+    UC_HOOK_MEM_WRITE_PROT_IDX,
+    UC_HOOK_MEM_FETCH_PROT_IDX,
+    UC_HOOK_MEM_READ_IDX,
+    UC_HOOK_MEM_WRITE_IDX,
+    UC_HOOK_MEM_FETCH_IDX,
+
+    UC_HOOK_MAX,
+};
+
+// for loop macros to loop over hook lists
+#define uc_hook_loop_common(uc, hh, idx, extra)                            \
+    struct list_item *cur;                                                 \
+    for (                                                                  \
+    cur = (uc)->hook[idx##_IDX].head;                                      \
+    cur != NULL && ((hh) = (struct hook *)cur->data)                       \
+        && !uc->stop_request /* stop excuting callbacks on stop request */ \
+        extra;               /* extra is used below for address ranges */  \
+    cur = cur->next)
+
+// loop without address check
+#define uc_hook_loop(uc, hh, idx) uc_hook_loop_common(uc, hh, idx, )
```

Comment:
for consistency, what do you think about renaming these macro to use uppercase instead? i was a bit confused when i first saw this.

for the loop macro, should we use more traditional name, such as HOOK_FOREACH?
(Qemu is using macros such as CPU_FOREACH, QTAILQ_FOREACH etc)